src/generic/eigen_solver.h

```diff
@@ -74,28 +74,92 @@ namespace oomph
     /// Empty destructor
     virtual ~EigenSolver() {}
 
-    /// Actual eigensolver. This takes a pointer to a problem and returns
+    /// Eigensolver. This takes a pointer to a problem and returns
     /// a vector of complex numbers representing the eigenvalues
-    /// and a corresponding vector of eigenvectors
-    virtual void solve_eigenproblem(Problem* const& problem_pt,
-                                    const int& n_eval,
-                                    Vector<std::complex<double>>& eigenvalue,
-                                    Vector<DoubleVector>& eigenvector) = 0;
-
+    /// and a corresponding vector of eigenvectors. n_eval specifies the min.
+    /// number of eigenvalues/vectors required. This is primarily used in
+    /// Arnoldi type implementations; direct solvers such as QZ compute all the
+    /// eigenvalues/vectors.
+    /// Note: this is a legacy version of this function that stores re & imag
+    /// parts of eigenvectors in some solver-specific collection of real
+    /// vectors.
+    virtual void solve_eigenproblem_legacy(
+      Problem* const& problem_pt,
+      const int& n_eval,
+      Vector<std::complex<double>>& eigenvalue,
+      Vector<DoubleVector>& eigenvector) = 0;
+
+
+    /// Solve the real eigenproblem that is assembled by elements in
+    /// a mesh in a Problem object. Note that the assembled matrices include the
+    /// shift and are real. The eigenvalues and eigenvectors are,
+    /// in general, complex, and the eigenvalues can be NaNs or Infs.
+    /// This function is therefore merely provided as a convenience, to be
+    /// used if the user is confident that NaNs don't arise (e.g. in Arnoldi
+    /// based solvers where typically only a small number of (finite)
+    /// eigenvalues are computed), or if the users is happy to deal with NaNs in
+    /// the subsequent post-processing.
+    /// Function is virtual so it can be overloaded for arnoldi type solvers
+    /// that compute the (finite) eigenvalues directly
+    /// At least n_eval eigenvalues are computed.
+    virtual void solve_eigenproblem(
+      Problem* const& problem_pt,
+      const int& n_eval,
+      Vector<std::complex<double>>& eigenvalue,
+      Vector<Vector<std::complex<double>>>& eigenvector)
```

Comment:
This interface is not suitable for  distributed Problems.. Do we need a complex DoubleVector class?

Reply:
I'll investigate.

Reply:
Rewrite to DoubleVectors done. The new eigensolvers now return two DoubleVectors, one for the real and one for the imaginary part of the actual, in general complex, eigenvectors.

Done